File: use_cases/inference_cpp/single_io_single_batch/single_io_single_batch.cpp
#include <iostream>
#include <set>
#include <memory>
#include <chrono>
#include <iomanip>

#include "nn_handler_lib/nn_handler.hpp"
#include "aux_fcn.hpp"

// PARAMETERS
/** \brief Number of inferences for warmup */
constexpr int n_inferences_warmup = 10;
/** \brief Number of inferences to calculate the average time */
constexpr int n_inferences = 100;
/** \brief Path of the ONNX model*/
const std::string path_model_onnx = "../../models/single_io/model_single_batch.onnx";
/** \brief Path to save the TensorRT engine for inference*/
const std::string path_engine_save = "../../models/single_io";
// Precision of the NN. Either FP16 or FP32
Precision precision = Precision::FP16;
// DLA core to use. If -1, it is not used
int dla_core = -1;
// GPU index (ORIN only has the 0 index)
int device_index=0;
// Batch size. If the model has fixed batch, this has to be 1. If the model has dynamic batch, this can be >1.
int batch_size = 1;
         
// Input and output files
const std::string input_path = "../../test_data/single_io/x.txt";
const std::string output_path = "../../test_data/single_io/y.txt";

int main(){

    // Create variable for inference
    NNHandler nn_handler(path_model_onnx, path_engine_save, precision, dla_core, device_index, batch_size);
    // Print the data of the handler
    nn_handler.print_data();

    // Load input and output ground truth. We will load all the batches but will only use the first one
    std::vector<std::vector<float>> input_all_batches = read_file(input_path);
    std::vector<std::vector<float>> output_gt_all_batches = read_file(output_path);

    // We only select the first batch for this experiment
    std::vector<float> input = input_all_batches[0];
    std::vector<float> output_gt = output_gt_all_batches[0];

    // Predicted output
    std::vector<float> output_pred;
    
    // Perform WARMUP inference (only with the first batch)
    for (int i=0; i< n_inferences_warmup; i++){
        nn_handler.run_inference(input, output_pred);
    }
    
    // Get the current time before inference
    auto start = std::chrono::high_resolution_clock::now();
    // Measure time of inference
    for (int i=0; i<n_inferences; i++){
        nn_handler.run_inference(input, output_pred);
    }
    // Get the current time after inference
    auto end = std::chrono::high_resolution_clock::now();

    // Calculate the duration in milliseconds
    std::chrono::duration<double, std::milli> duration = end - start;
    std::cout << "Time taken to perform inference: " << duration.count()/n_inferences << " milliseconds" << std::endl;

    // Show the results
    std::cout << std::fixed << std::setprecision(6);
    std::cout << "Ground truth output: " << std::endl;
    for (int i=0; i<10; i++){
        std::cout << output_gt[i] << " ";
    }
    std::cout << "\nPredicted output: " << std::endl;
    for (int i=0; i<10; i++){
        std::cout << output_pred[i] << " ";
    }
    std::cout << std::endl;

    std::cout << "Mean square error: " << calculate_mae(output_gt, output_pred, 10) << std::endl;

    return 0;
}

File: use_cases/inference_python/single_io_single_batch.py

import numpy as np
import sys
import os
import tensorrt as trt
import time

engine_path = os.path.abspath(os.path.join("..", "..", "tensorrt_python" ))
sys.path.insert(0, engine_path)
from tensorrt_engine import TensorRTEngine, Options

# PARAMETERS
# Number of inferences for warmup
n_inferences_warmup = 10
# Number of inferences to calculate the average time */
n_inferences = 100
# Path of the ONNX model*/
path_model_onnx = "../models/single_io/model_single_batch.onnx"
# Path to save the TensorRT engine for inference*/
path_engine_save = "../models/single_io"
# Precision of the NN. Either FP16 or FP32
precision = trt.float16
# DLA core to use. If -1, it is not used
dla_core = -1
# GPU index (ORIN only has the 0 index)
device_index=0
# Batch size. If the model has fixed batch, this has to be 1. If the model has dynamic batch, this can be >1.
batch_size = 1

# Input and output files
input_path = "../test_data/single_io/x.txt"
output_path = "../test_data/single_io/y.txt"

if __name__ == "__main__":

    # We create the options variable, and only modify the out_file_path. We leave the rest of options by default
    trt_options = Options()
    trt_options.out_file_path = path_engine_save
    trt_options.precision = precision
    trt_options.batch_size = batch_size
    trt_options.deviceIndex = device_index
    trt_options.dlaCore = dla_core
    trt_options.out_file_path = path_engine_save

    # Create the engine, build and load the network
    engine = TensorRTEngine(trt_options)
    engine.build(path_model_onnx)
    engine.loadNetwork()

    # Print engine data
    engine.print_data()

    # Get the dimensions from the engine
    input_dims = engine.input_dims[0]
    output_dims = engine.output_dims[0]


    # Load data
    input_all_batches = np.loadtxt(input_path, delimiter=',')
    output_gt_all_batches = np.loadtxt(output_path, delimiter=',')

    # Take first batch
    input = input_all_batches[0,:]
    output_gt = output_gt_all_batches[0,:]

    # Reshape to meet dimensions
    input = input.reshape(input_dims)
    output_gt = output_gt.reshape(output_dims)

    # The function expects a list of inputs
    inputs = [input]

    # Warmup inference
    for i in range(n_inferences_warmup):
        _ = engine.do_inference_v2(inputs)

    # Measure inference time
    start_time = time.time()
    for i in range(n_inferences):
        output_pred = engine.do_inference_v2(inputs)
    end_time = time.time()
    avg_inference_time = (end_time - start_time)*1000.0 / n_inferences
    print(f"Average inference time: {avg_inference_time:.6f} miliseconds")

    # Show results
    print("Ground truth output:\n", output_gt)
    print("Predicted output:\n", output_pred[0])

    # Show errors
    mae_output = np.mean(np.abs(output_gt - output_pred[0]))
    print("Mean square error: ", mae_output)
